### PR TITLE
Feature/csi 4165 host definer server delete

### PR DESCRIPTION
**Pull request by matancarmeli7**

### CODE REVIEW

File: controllers/servers/host_definer/storage_manager/host_definer_server.py [hunk 1]
```diff
@@ -44,4 +44,21 @@ def define_host(self, request):  # pylint: disable=invalid-name
             return DefineHostResponse(error_message=str(ex))
 
     def undefine_host(self, request):  # pylint: disable=invalid-name
-        raise NotImplementedError
+        node_id_info = NodeIdInfo(request.node_id)
+        initiators = node_id_info.initiators
+        try:
+            array_connection_info = get_array_connection_info_from_secrets(request.system_info)
+            array_type = detect_array_type(array_connection_info.array_addresses)
```
Comment (oriyarde) [inline controllers/servers/host_definer/storage_manager/host_definer_server.py:51]: can't we "cache" this in the k8s host definition resource?

File: controllers/servers/host_definer/storage_manager/host_definer_server.py [hunk 2]
```diff
@@ -44,4 +44,21 @@ def define_host(self, request):  # pylint: disable=invalid-name
...
+            with get_agent(array_connection_info, array_type).get_mediator() as array_mediator:
+
+                try:
+                    found_host_name, _ = array_mediator.get_host_by_host_identifiers(initiators)
+                    logger.debug("host found : {}".format(found_host_name))
+                    array_mediator.delete_host(found_host_name)
+                except HostNotFoundError:
+                    logger.debug("host was not found")
+
+                return DefineHostResponse()
```
Comment (oriyarde) [inline controllers/servers/host_definer/storage_manager/host_definer_server.py:61]:
```suggestion
                return UndefineHostResponse()
```

File: controllers/tests/controller_server/host_definer_server_test.py [hunk 3]
```diff
@@ -32,17 +32,20 @@ def setUp(self):
...
+        self.request.node_id = '{};;;{}'.format(self.hostname, self.iqn)
+        self.request.system_info = {'username': 'user', 'password': 'pass', 'management_address': 'mg111'}
+
+
+class TestDefineHost(BaseSetUp):
 
     def _prepare_define_host_success(self, is_idempotency=False):
         if is_idempotency:
-            self.mediator.get_host_by_host_identifiers.return_value = ("found_host_name", "")
+            self.mediator.get_host_by_host_identifiers.return_value = (VOLUME_NAME, '')
```
Comment (oriyarde) [inline controllers/tests/controller_server/host_definer_server_test.py:48]:
```suggestion
            self.mediator.get_host_by_host_identifiers.return_value = (HOST_NAME, '')

```

**Review by oriyarde — DISMISSED** [review]

**Review by oriyarde — APPROVED** [review]

File: controllers/servers/host_definer/storage_manager/host_definer_server.py [hunk 4]
```diff
@@ -44,4 +44,21 @@ def define_host(self, request):  # pylint: disable=invalid-name
...
+        node_id_info = NodeIdInfo(request.node_id)
+        initiators = node_id_info.initiators
+        try:
+            array_connection_info = get_array_connection_info_from_secrets(request.system_info)
+            array_type = detect_array_type(array_connection_info.array_addresses)
+            with get_agent(array_connection_info, array_type).get_mediator() as array_mediator:
+
+                try:
+                    found_host_name, _ = array_mediator.get_host_by_host_identifiers(initiators)
+                    logger.debug("host found : {}".format(found_host_name))
```
Comment (matancarmeli7) [inline controllers/servers/host_definer/storage_manager/host_definer_server.py:56]: like [this comment](https://github.com/IBM/ibm-block-csi-driver/pull/535#discussion_r930697235)

Comment (ArbelNathan) [reply controllers/servers/host_definer/storage_manager/host_definer_server.py:56]: done https://github.com/IBM/ibm-block-csi-driver/pull/560

File: controllers/servers/host_definer/storage_manager/host_definer_server.py (hunk 1 above)
Comment (ArbelNathan) [reply controllers/servers/host_definer/storage_manager/host_definer_server.py:51]: @matancarmeli7 what do you think?

Comment (ArbelNathan) [reply controllers/servers/host_definer/storage_manager/host_definer_server.py:51]: CSI-4603

File: controllers/tests/controller_server/host_definer_server_test.py (hunk 3 above)
Comment (ArbelNathan) [reply controllers/tests/controller_server/host_definer_server_test.py:48]: done https://github.com/IBM/ibm-block-csi-driver/pull/560

File: controllers/servers/host_definer/storage_manager/host_definer_server.py (hunk 2 above)
Comment (ArbelNathan) [reply controllers/servers/host_definer/storage_manager/host_definer_server.py:61]: after talking with @matancarmeli7 -  we will do it only if any need to distinguish between the responses

**Pull request merged**